src/mappings/clearingHouse.ts

```diff
@@ -399,7 +399,7 @@ export function handleReferralPositionChanged(event: ReferredPositionChanged): v
     let code = event.params.referralCode.toString()
 
     if (positionChangedEvent !== null) {
-        let trader = getOrCreateTrader(event.transaction.from)
+        let trader = getOrCreateTrader(positionChangedEvent.trader as Address)
```

Comment:
If trader is a contract, `event.transaction.from` is the caller of contract instead of contract itself. However, the contract address is the actual address which uses referral code, not its caller.